### PR TITLE
🎉 Add Source Oracle

### LINKED ISSUE #1232: Oracle database connector
## Tell us about the new integration you’d like to have
Source/Destination from/to oracle database

## Describe the context around this new integration
To do ETL

## Describe the alternative you are considering or using
Airflow

### LINKED ISSUE #1685: Oracle source connector
## Tell us about the new integration you’d like to have
Oracle 11g R2 and up, replication to Snowflake and/or snapshots to Azure Data Lake blob storage, hourly frequency or lower.

Since the source is a high-traffic ERP system, log-based data capture for incremental updates would be ideal to reduce load on the source.

## Describe the context around this new integration
This is (and incremental updates for SQL Server) is the main blocker for our org being an early adopter. Most of our standard enterprise-y applications are backed by either Oracle or mssql

## Describe the alternative you are considering or using
Hand-rolled or Azure Data Factory

Note that there's already a [Singer tap](https://github.com/singer-io/tap-oracle) ([docs](https://www.stitchdata.com/docs/integrations/databases/oracle)) that appears to also support log-based replication

### PR DESCRIPTION
**Pull request by marcosmarxm**
## What
Add new source Oracledb to resolves #1685 and resolves #1232

## How
Use the JBDC base code to add Oracledb as a source.

## Pre-merge Checklist
- [x] *SourceOracleTest*
- [x] *OracleJdbcStandardSourceTest*
- [x] *OracleStressTest*
- [x] *OracleSourceStandardTest*
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `OracleSource.java`
1. `OracleSourceTest.java`
1. `OracleStressTest.java`
2. `OracleJdbcStandardSourceTest.java`
3. `OracleSourceStandardTest.java`

### CODE REVIEW

**Comment by sherifnada** [comment]
@marcosmarxm let us know when you'd like a review! Feel free to assign me as a reviewer when you're ready

**Comment by marcosmarxm** [comment]
@sherifnada sure! Just changed the copy-paste code from Postgres source. After the tests work correctly, I request your review :rocket:

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceStandardTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        "org.oracle.Driver",
+        SQLDialect.POSTGRES);
+
+    database.query(ctx -> {
+      ctx.fetch("CREATE TABLE id_and_name(id INTEGER, name VARCHAR(200));");
+      ctx.fetch("INSERT INTO id_and_name (id, name) VALUES (1,'picard'),  (2, 'crusher'), (3, 'vash');");
+      ctx.fetch("CREATE TABLE starships(id INTEGER, name VARCHAR(200));");
+      ctx.fetch("INSERT INTO starships (id, name) VALUES (1,'enterprise-d'),  (2, 'defiant'), (3, 'yamato');");
+      return null;
+    });
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceStandardTest.java:85]: @sherifnada here is an example. `database.query` invoke the `transform.query(DSL.using(ds, dialect));` function. In the free version of jooq Oracle isn't available.

Comment (cgardens) [reply airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceStandardTest.java]: @marcosmarxm jooq's free version also doesn't support oracle. don't know if using `MssqlSourceStandardTest`as an example might help you out here as an example of how to do this without jooq.

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceStandardTest.java]: Thank @cgardens I changed the `.query` function to `.createStatement().execute` as *MssqlSourceStandardTest** does

**Comment by marcosmarxm** [comment]
Oracle Source Test and Integration Test are working :rocket:. 
The OracleJdbcStandardSourceTest gets messages equal expected but the assertions fail.

Some points about the Oracle implementation your opinion is valuable: 
- Oracle doesn't allow multiple inserts then I break inserts into individual ones, for other JDBC databases this is not a problem.
- Oracle doesn't allow `;` at the end of the line (for now I hard removed for ease);
- In the Dockerfile, I enter the env `TZ=UTC` because Oracle database needs this information, I didn't test if this could change values in DateTime response yet.
- It's very complicated to use `DROP TABLE` when you don't have the table created, the same is true when you execute a `CREATE TABLE` when you already had one.
- The records key name returned are in uppercase, this could be solved by applying a lower function:
```
Expected :[io.airbyte.protocol.models.AirbyteMessage@42205254[
type=RECORD,log=<null>,spec=<null>,connectionStatus=<null>,catalog<null>,
record=io.airbyte.protocol.models.AirbyteRecordMessage@48db1110[
stream=SYSTEM.ID_AND_NAME,data={"ID":1,"NAME":"picard","UPDATED_AT":"2004-10-19"},
emittedAt=<null>,additionalProperties={}],state=<null>,additionalProperties={}],

Actual   :[io.airbyte.protocol.models.AirbyteMessage@7ca1e9b1[
type=RECORD,log=<null>,spec=<null>,connectionStatus=<null>,catalog=<null>,
record=io.airbyte.protocol.models.AirbyteRecordMessage@6761adeb[
stream=SYSTEM.ID_AND_NAME,data={"ID":1,"NAME":"picard","UPDATED_AT":"2004-10-19"},
emittedAt=<null>,additionalProperties={}],state=<null>,additionalProperties={}],
```

**Comment by marcosmarxm** [comment]
Left is the expected value and right is the actual values. Look Oracle is returning the INTEGER column as a BigInteger. Oracle is creating the table as NUMBER not integer.
Screenshot from 2021-03-22 14-42-11

**Review by sherifnada — COMMENTED** [review]
Also commenting on issues you mentioned via Slack DM: 

`Oracle create table statement for INTEGER create as NUMBER and when someone query the return become BigDecimal. This doesnt match the Integer type expected;`
This is fine to change in JDBC tests IMO as type-specific tests should live in each DB's test suite

`Oracle raise errors for CREATE/DROP if not exists or already exist;`
is this possible to handle in the test? I would rather that we don’t entirely hide these errors in case there is an actual problem in the code. Also I’m not sure the error codes you are ignoring are generic for all SQL databases.

File: airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java [hunk 2]
```diff
@@ -47,7 +47,47 @@
...
   default void execute(String sql) throws SQLException {
-    execute(connection -> connection.createStatement().execute(sql));
+    try {
+      execute(connection -> connection.createStatement().execute(sql));
+    } catch (SQLException e) {
+      printSQLException(e);
+    }
+  }
+
+  static void printSQLException(SQLException ex) {
```
Comment (sherifnada) [inline airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java:57]: I am worried about swallowing exceptions wholesale like this. Can we instead ensure the problem isn't happening in the query itself?

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+          printSQLException(e);
+      }
+  }
+
+  public static void printSQLException(SQLException ex) {
+        for (Throwable e : ex) {
+            if (e instanceof SQLException) {
+                if (ignoreSQLException(((SQLException)e).getSQLState()) == false) {
+                    e.printStackTrace(System.err);
+                    System.err.println("SQLState: " + ((SQLException)e).getSQLState());
```
[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java:135]: I'll change this, I just copy-pasta code from the Oracle example.

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java [hunk 4]
```diff
@@ -603,49 +616,30 @@ private void incrementalCursorCheck(
...
 
   // get catalog and perform a defensive copy.
-  private static ConfiguredAirbyteCatalog getConfiguredCatalogWithOneStream(final String defaultNamespace) {
-    final ConfiguredAirbyteCatalog catalog = CatalogHelpers.toDefaultConfiguredCatalog(getCatalog(defaultNamespace));
-    // Filter to only keep the main stream name as configured stream
-    catalog.withStreams(catalog.getStreams().stream().filter(s -> s.getStream().getName().equals(streamName)).collect(Collectors.toList()));
-    return catalog;
-  }
-
-  private static AirbyteCatalog getCatalog(final String defaultNamespace) {
```
[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:613]: Yes, my JdbcSourceStandardTest version is not updated. I'll get latest from master branch.

File: airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java [hunk 5]
```diff
@@ -47,7 +47,47 @@
...
+
+  static boolean ignoreSQLException(String sqlState) {
+    if (sqlState == null) {
+      System.out.println("The SQL state is not defined!");
+      return false;
+    }
+    // X0Y32: Jar file already exists in schema
+    if (sqlState.equalsIgnoreCase("X0Y32"))
+      return true;
+    // 42Y55: Table already exists in schema
```
[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java:83]: I think this is specific to Oracle, but I'll test for other databases using JDBC driver.

File: airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java:57]: For Oracle no, the problem is not the query. This is to handle statements like `DROP TABLE XPTO` and the table `XPTO` doesn't exist in Oracle.
```
SQLState: 42Y55
Error Code: 30000
Message: 'DROP TABLE' cannot be performed on
'TESTDB.XTPO' because it does not exist.
```

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java [hunk 6]
```diff
@@ -83,18 +85,31 @@
...
+  public static String COL_NAME = "name";
+  public static String COL_UPDATED_AT = "updated_at";
+  public String COL_FIRST_NAME = "first_name";
+  public String COL_LAST_NAME = "last_name";
+  public String COL_LAST_NAME_WITH_SPACE = "last name";
+  public Number COL_ID_VALUE_1 = 1;
+  public Number COL_ID_VALUE_2 = 2;
+  public Number COL_ID_VALUE_3 = 3;
+  public Number COL_ID_VALUE_4 = 4;
+  public Number COL_ID_VALUE_5 = 5;
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:112]: @sherifnada instead of overriding the function, I replace the hard-coded values with variables. Then, I just replace them in the Oracle class.

File: airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcDatabase.java]: I removed this from JdbcDatabase class.

**Comment by marcosmarxm** [comment]
Almost all tests from JdbcSourceStandardTest are OK with some small workaround! :laughing: 

### Oracle doesn't support Integer
_tl;dr problem_: in `setup()` is created a table `CREATE TABLE id_and_name(id INTEGER)` this work for some databases. But Oracle will save INTEGER as NUMERIC type. When I tried to query this data and compare it with expected records tests will fail. Because the **expected** record is Integer type (works for Postgres, Mssql, etc) but **actual** records returned are mapped to BigDecimal. 

I made a workaround creating variables to numeric types... but in `testReadMultipleTables` expected messages are mapped to Json. BigDecimal values are converted to Integer, this breaks the comparison between expected vs actual messages from Oracle. 

I added a line to workaround this `((ObjectNode) m.getRecord().getData()).replace(COL_ID, Jsons.jsonNode(BigDecimal.valueOf(m.getRecord().getData().get(COL_ID).asInt())));`  This will break for other databases, I only added to discuss a solution.
https://github.com/airbytehq/airbyte/blob/075502edead0a5e99d5513f132e7c95f96a8e5e8/airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java#L337-L347

About this problem returning BigDecimal and not Integer, I talked to @davinchia: https://www.jooq.org/doc/3.14/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-zero-scale-decimals/

> A zero-scale decimal, such as DECIMAL(10) or NUMBER(10, 0) is really an integer type with a decimal precision rather than a binary / bitwise precision. Some databases **(e.g. Oracle) do not support actual integer types at all, only decimal types.** Historically, jOOQ generates the most appropriate integer wrapper type instead of BigDecimal or BigInteger:
> 
> NUMBER(2, 0) and less: java.lang.Byte
> NUMBER(4, 0) and less: java.lang.Short
> NUMBER(9, 0) and less: java.lang.Integer
> NUMBER(19, 0) and less: java.lang.Long

Airbyte at this moment converts NUMERIC types to BigDecimal. This pattern match should consider the Oracle case with scale and precision?
https://github.com/airbytehq/airbyte/blob/794796cbd0bd545e3ac83ae322aec45b0835d8c2/airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcUtils.java#L190

Suggestions? @sherifnada @ChristopheDuong

**Comment by marcosmarxm** [comment]
### Oracle doesn't support `;` at the end of the statement
Other databases run queries without the `;` at the end of the statement. Could be a problem in the future if someone adds new tests with `;`. Is there a way to prevent this?

**Comment by cgardens** [comment]
woof! thanks for drilling into this.

this is where trying to reuse the same tests for all databases bites us a bit. i haven't dug super deeply into all of the details you mentioned, but i think pretty much we are going to have to commit some sin to make this work. we already do this [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java#L233-L236) where internal to the tests we have a slightly different behavior based on the database. 

i think we probably just need to do something like that for the oracle db. i guess i would ask, what's the smallest conditional that switches on whether the db being used is oracle that you can add get the tests to work for oracle? does that help? i can dig in deeper on this if it would be helpful to try to answer that question, but you probably have enough context to propose something already.

File: airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcUtils.java [hunk 7]
```diff
@@ -136,7 +131,8 @@ private static void setJsonField(ResultSet r, int i, ObjectNode o) throws SQLExc
       case BIGINT -> o.put(columnName, nullIfInvalid(() -> r.getLong(i)));
       case FLOAT, DOUBLE -> o.put(columnName, nullIfInvalid(() -> r.getDouble(i), Double::isFinite));
       case REAL -> o.put(columnName, nullIfInvalid(() -> r.getFloat(i), Float::isFinite));
-      case NUMERIC, DECIMAL -> o.put(columnName, nullIfInvalid(() -> r.getBigDecimal(i)));
+      case NUMERIC -> o.put(columnName, nullIfInvalid(() -> r.getInt(i)));
+      case DECIMAL -> o.put(columnName, nullIfInvalid(() -> r.getBigDecimal(i)));
```
Comment (marcosmarxm) [inline airbyte-db/src/main/java/io/airbyte/db/jdbc/JdbcUtils.java:135]: @cgardens this way I don't need to change the `JdbcStandardTest`. Can you validate this change? I can't measure the impact of this change.

**Comment by marcosmarxm** [comment]
The current code works with Oracle and Postgres :traffic_light:. For databases without schemas it will break, I need to change how tables are transferred to SQL statements.

I only need to add this Oracle condition to make things works
https://github.com/airbytehq/airbyte/blob/30243e0e88a44bc023120e566325d126ffdeb2b3/airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java#L162-L164

**Comment by marcosmarxm** [comment]
@cgardens I added the helper method to convert int to BigDecimal when the driver is Oracle. This way JdbcUtils don't be changed. I apply tests for Oracle, Postgres and also Mssql (this one doesn't support schema) all 3 passed.

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java [hunk 8]
```diff
@@ -31,11 +31,11 @@
 import io.airbyte.integrations.source.jdbc.AbstractJdbcSource;
 import io.airbyte.integrations.source.jdbc.test.JdbcSourceStandardTest;
 import java.math.BigDecimal;
+import java.sql.*;
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java:34]: we tend to avoid using the `.*` syntax. there's a setting in intellij that lets you set it very high so it doesn't kick in.

Screen Shot 2021-03-30 at 4 20 32 PM

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java [hunk 9]
```diff
@@ -145,31 +163,58 @@ public void setup() throws Exception {
...
-          .execute(String.format("CREATE TABLE %s(id INTEGER, name VARCHAR(200), updated_at DATE, PRIMARY KEY (id));",
-              getFullyQualifiedTableName(TABLE_NAME)));
+
       connection.createStatement().execute(
           String.format(
-              "INSERT INTO %s(id, name, updated_at) VALUES (1,'picard', '2004-10-19'),  (2, 'crusher', '2005-10-19'), (3, 'vash', '2006-10-19');",
+              "CREATE TABLE %s(id INTEGER, name VARCHAR(200), updated_at DATE, PRIMARY KEY (id))",
+              getFullyQualifiedTableName(TABLE_NAME)));
+      connection.createStatement().execute(
+          String.format("INSERT INTO %s(id, name, updated_at) VALUES (1,'picard', '2004-10-19')",
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:179]: does oracle db not support multi line inserts? we can't just remove the semi colons here? we need to split it into 4 separate inserts?

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java [hunk 10]
```diff
@@ -84,18 +85,35 @@
...
-  private static final String TABLE_NAME_WITHOUT_PK = "id_and_name_without_pk";
-  private static final String TABLE_NAME_COMPOSITE_PK = "full_name_composite_pk";
-
-  private JsonNode config;
-  private JdbcDatabase database;
-  private AbstractJdbcSource source;
-  private static String streamName;
+  public static String SCHEMA_NAME = "jdbc_integration_test1";
+  public static String SCHEMA_NAME2 = "jdbc_integration_test2";
+  public static String SCHEMA_NAME3 = "jdbc_integration_test3";
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:90]: why do we need so many different schema names? we only had 2 before. i'm a bit confused how the extra ones are being used.

File: airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+      "port": {
+        "title": "Port",
+        "description": "Port of the database.",
+        "type": "integer",
+        "minimum": 0,
+        "maximum": 65536,
+        "default": 1521,
+        "examples": ["1521"]
+      },
+      "sid": {
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json:24]: i'm not really familiar with what this is. should it be a required field?

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.containers.OracleContainer;
+
+class OracleSourceTest {
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java:57]: do we need this test class at all? i think we can skip it unless you feel it is testing something important.

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.OracleContainer;
+
+/**
+ * Runs the stress tests in the source-jdbc test module. We want this module to run these tests
+ * itself as a sanity check. The trade off here is that this class is duplicated from the one used
+ * in source-postgres.
+ */
+@Disabled
+class OracleStressTest extends JdbcStressTest {
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java:52]: does this test actually pass? just want to sanity check.

**Review by cgardens — COMMENTED** [review]
looks good! a few last questions but i think we're almost there!

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java (hunk 9 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:179]: Oracle doesn't support multi-insert. Unfortunately, we need to break into individual inserts and remove semi-colons to be compatible Oracle with others databases.
https://stackoverflow.com/questions/39576/best-way-to-do-multi-row-insert-in-oracle
Another example from [Oracle tutorial](https://livesql.oracle.com/apex/livesql/file/content_BM1LJQ87M5CNIOKPOWPV6ZGR3.html)

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java (hunk 13 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java:52]: I copied it from Postgresql and kept it, because is disabled will not run anyway, so..

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java (hunk 12 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java:57]: It's very basic testing and JdbcStandardTest has the test `testReadSuccess` very similar to the one applied here but more comprehensive.

File: airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json (hunk 11 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json:24]: It's similar to the database name in Postgres.
SID = the unique name of your DB,

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java (hunk 9 above)
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java:179]: cool. thanks for explaining.

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java (hunk 13 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleStressTest.java:52]: I read wrong and totally mislead myself here. No tests are not passing here.

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java (hunk 10 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceStandardTest.java]: I created a function called `cleanUpTables` and remove all extra schemas.

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java (hunk 12 above)
Comment (cgardens) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java:57]: let's kill it. i think we added this test in postgres to test a specific character encoding case. i think we test this stuff well enough in the jdbc standard tests for oracle.

File: airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java (hunk 8 above)
Comment (cgardens) [reply airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcStandardSourceTest.java]: i'm not sure it does. it's an added style decision we made on top of the google java style

File: airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json (hunk 11 above)
Comment (cgardens) [reply airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json:24]: should it be required? or is it okay if it is left blank?

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json:24]: It's required

**Review by cgardens — APPROVED** [review]

File: airbyte-integrations/connectors/source-oracle/build.gradle [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    id 'airbyte-integration-test-java'
+}
+
+application {
+    mainClass = 'io.airbyte.integrations.source.oracle.OracleSource'
+}
+
+dependencies {
+
+    implementation group: 'xerces', name: 'xercesImpl', version: '2.12.1'
```
Comment (cgardens) [inline airbyte-integrations/connectors/source-oracle/build.gradle:13]:
```suggestion
    // required so that log4j uses a standard xml parser instead of an oracle one (that gets pulled in by the oracle driver)
    implementation group: 'xerces', name: 'xercesImpl', version: '2.12.1'
```
